### PR TITLE
Make and test some small improvements to `lazy`

### PR DESCRIPTION
**Pull request by wchargin**
Follow-up to #1781. See individual commit messages for details. To be
merged with rebase.

### CODE REVIEW

File: tensorboard/lazy.py [hunk 1]
```diff
@@ -44,10 +44,14 @@ def wrapper(load_fn):
     # make future lookups efficient (only failed lookups call __getattr__).
     @_memoize
     def load_once(self):
+      if load_once.loading:
+        raise ImportError("Circular import when resolving LazyModule %r" % name)
```
Comment (nfelt) [inline tensorboard/lazy.py]: Optional but seems like you might as well put this logic in `_memoize` itself.

File: tensorboard/lazy_test.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import unittest
+
+from tensorboard import lazy
+
+
+class LazyTest(unittest.TestCase):
+  """Tests the TensorBoard program."""
```
Comment (nfelt) [inline tensorboard/lazy_test.py]: Update comment?

File: tensorboard/lazy_test.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    def inner():
+      return outer.foo
+
+    @lazy.lazy_load("outer")
+    def outer():
+      return inner
+
+    with self.assertRaises(ImportError) as cm:
+      outer.bar
+    self.assertEqual(
```
Comment (nfelt) [inline tensorboard/lazy_test.py]: six.assertRaisesRegex()?

File: tensorboard/lazy_test.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        str(cm.exception),
+        "Circular import when resolving LazyModule 'inner'",
+    )
+
+  def test_repr_before_load(self):
+    @lazy.lazy_load("foo")
+    def foo():
+      raise ValueError("Should not need to resolve this module.")
```
Comment (nfelt) [inline tensorboard/lazy_test.py]: IMO `self.fail("...")` slightly more idiomatic for this

File: tensorboard/lazy.py (hunk 1 above)
Comment (wchargin) [reply tensorboard/lazy.py]: But `_memoize` takes a function that accepts a single hashable argument,
so recursion is perfectly fine. If `_memoize` took a niladic function,
the argument would be more convincing, I think.

File: tensorboard/lazy_test.py (hunk 2 above)
Comment (wchargin) [reply tensorboard/lazy_test.py]: Yes; thanks.

File: tensorboard/lazy_test.py (hunk 3 above)
Comment (wchargin) [reply tensorboard/lazy_test.py]: Sure; will change.

File: tensorboard/lazy_test.py (hunk 4 above)
Comment (wchargin) [reply tensorboard/lazy_test.py]: Agreed; will change.

File: tensorboard/lazy.py (hunk 1 above)
Comment (wchargin) [reply tensorboard/lazy.py]: If you don’t mind, I’ll keep it as written in this PR, because that
enables us to give a better error message.

**Pull request merged**